text/0225-configuration.md

```diff
@@ -0,0 +1,335 @@
+# OpenTelemetry Configuration
+
+A new configuration interface is proposed here in the form of a configuration model, which can be expressed as a file, and validated through a published schema.
+
+## Motivation
+
+OpenTelemetry specifies code that can operate in a variety of ways based on the end-user’s desired mode of operation. This requires a configuration interface be provided to the user so they are able to communicate this information. Currently, OpenTelemetry specifies this interface in the form of the API exposed by the SDKs and environment variables. This environment variable interface is limited in the structure of information it can communicate and the primitives it can support.
+
+### Environment Variable Interface Limitations
+
+The environment variable interface suffers from the following identified limitations:
+
+* **Flat**. Structured data is only allowed by using higher-level naming or data encoding schemes. Examples of configuration limited by lack of structured configuration include:
+  * Configuring multiple span processors, periodic metric readers, or log record processors.
+  * Configuring views.
+  * Configuring arguments for parent based sampler (sampler parent is remote and sampled vs. not sampled, sampler when parent is local and sampled vs. not sampled).
+* **Runtime dependent**. Different systems expose this interface differently (Linux, BSD, Windows). This usually means unique instructions are required to properly interact with the configuration interface on different systems.
+* **Limited values**. Many systems only allow string values to be used, but OpenTelemetry specifies many configuration values other than this type. For example, OTEL_RESOURCE_ATTRIBUTES specifies a list of key value pairs to be used as resource attributes, but there is no way to specify array values, or indicate that the value should be interpreted as non-string type.
+* **Limited validation**. Validation can only be performed by the receiver, there is no meta-configuration language to validate input.
+* **Difficult to extend**. It’s difficult to anticipate the requirements of configuring custom extension components (processors, exporters, samplers, etc), and likely not practical to represent them in a flat structure. As a result, the environment variable interface is limited to components explicitly mentioned in the specification.
+* **Lacks versioning**. The lack of versioning support for environment variables prevents evolution over time.
+
+## Explanation
+
+Using a configuration model or configuration file, users can configure all options currently available via environment variables.
+
+### Goals
```

Comment:
I have a (rather strong) opinion that setting set via env vars has higher priority (takes precedence) over a setting set via configiuration file and it should be marked as a goal.

Reply:
Can you elaborate on what you mean by higher priority? 

Maybe you mean that environment variables should somehow be "layered" on top of a configuration file? If so, I disagree. Consider a config file that says something like:
```
sdk:
  tracer_provider:
    exporters:
      otlp/a:
        endpoint: http://endpoint1
      otlp/b:
        endpoint: http://endpoint2
    span_processors:
      - name: batch
         args:
           exporter: otlp/a
      - name: batch
         args:
           exporter: otlp/b
```
And consider that the environment variable is set with `OTEL_EXPORTER_OTLP_ENDPOINT=http://endpoint3`. How would the environment variable take higher priority over the file config? If you replace the endpoints of the two exporters in the file config with the env variable value `http://endpoint3`, you're almost certainly not doing what the user would want. 

Any scheme where environment variables have priority over a config file will require some sort of standard mapping between the environment variable schema and file config scheme. IMO, its impossible to define such a mapping which is intuitive in all cases, so better not to try. 

Nothing is forcing users to use file based config - its opt in. If they do opt in, they're opting into the documented behavior in which the config file represents the source of truth for configuration. If they wish to customize the experience with additional layers / overrides, they have a couple of tools:
1. They can use the fact the `Configure(config)` API accepts a config model as an argument, and provide their own customizations to the model after the initial parsing of the file via `Parse(file)`. An example of such a customization would be to interpret environment variables and apply them to the model in a way they decide makes sense.
2. They can use environment variable substitution to reference environment variables directly in a configuration file.

This is a pretty flexible toolkit, and allows users to have simple, intuitive out of the box behavior, while allowing for customization if the default is insufficient.

Reply:
I agree with you 👍 😉